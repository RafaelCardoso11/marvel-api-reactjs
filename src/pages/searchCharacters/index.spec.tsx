import RoutersContextTesting from "@/tests/routers";
import { SearchCharacters } from ".";
import { fireEvent, render, within } from "@testing-library/react";
import { Route } from "react-router-dom";

describe("<SearchCharacters />", () => {
  afterEach(() => {
    history.pushState({}, '', '/')
  })
  it("Deveria renderizar na tela os containers SearchCharactersPage, SearchCharactersPageTitle, SearchCharactersTitleIcon, SearchCharactersPageSearchContainer, SearchCharactersPageFooter", () => {
    const screen = render(
      <RoutersContextTesting>
        <SearchCharacters />
      </RoutersContextTesting>
    );

    const SearchCharactersPage = screen.getByTestId(
      "id-search-characters-page"
    );
    const SearchCharactersPageTitle = screen.getByTestId(
      "id-search-characters-title"
    );
    const SearchCharactersTitleIcon = screen.getByTestId(
      "id-search-characters-title-icon"
    );
    const SearchCharactersTitleIconContainer = screen.getByTestId(
      "id-search-characters-title-icon-container"
    );
    const SearchCharactersPageSubtitle = screen.getByTestId(
      "id-search-characters-subtitle"
    );
    const SearchCharactersPageSearchContainerBar = screen.getByTestId(
      "id-search-characters-container"
    );
    const SearchCharactersPageSearchBar = screen.getByTestId(
      "id-search-characters-bar"
    );
    const SearchCharactersPageSearchButton = screen.getByTestId(
      "id-search-characters-button"
    );
    const SearchCharactersPageFooter = screen.getByTestId(
      "id-search-characters-page-footer"
    );

    expect(SearchCharactersPage).toBeInTheDocument();
    expect(SearchCharactersPageTitle).toBeInTheDocument();
    expect(SearchCharactersTitleIconContainer).toBeInTheDocument();
    expect(SearchCharactersTitleIcon).toBeInTheDocument();
    expect(SearchCharactersPageSubtitle).toBeInTheDocument();
    expect(SearchCharactersPageSearchContainerBar).toBeInTheDocument();
    expect(SearchCharactersPageSearchBar).toBeInTheDocument();
    expect(SearchCharactersPageSearchButton).toBeInTheDocument();
    expect(SearchCharactersPageFooter).toBeInTheDocument();

    expect(screen.container).toMatchSnapshot();
    expect.assertions(10);
  });
  it("Deveria conter no titulo o Texto 'Search for your favorite hero e um Icon de Search'", () => {
    const screen = render(
      <RoutersContextTesting>
        <SearchCharacters />
      </RoutersContextTesting>
    );

    const SearchCharactersPageTitle = screen.getByTestId(
      "id-search-characters-title"
    );

    const iconInTitle = within(SearchCharactersPageTitle).getByTestId(
      "id-search-characters-title-icon"
    );
    expect(SearchCharactersPageTitle).toHaveTextContent(
      /Search for your favorite hero/i
    );
    expect(iconInTitle).toBeInTheDocument();
  });
  it("Deveria conter em tela um subtitle com o texto 'Find information about your favorite Marvel heroes.'", () => {
    const screen = render(
      <RoutersContextTesting>
        <SearchCharacters />
      </RoutersContextTesting>
    );

    const SearchCharactersPageSubTitle = screen.getByTestId(
      "id-search-characters-subtitle"
    );
    expect(SearchCharactersPageSubTitle).toHaveTextContent(
      /Find information about your favorite Marvel heroes./i
    );
  });
  it("Deveria conter em tela um Footer com o texto '⚡️ Não julgue o livro pela capa! O design pode não ser de outro mundo, mas o código é digno do Multiverso! ✨🚀'", () => {
    const screen = render(
      <RoutersContextTesting>
        <SearchCharacters />
      </RoutersContextTesting>
    );

    const SearchCharactersPageFooter = screen.getByTestId(
      "id-search-characters-page-footer"
    );
    expect(SearchCharactersPageFooter).toHaveTextContent(
      /⚡️ Não julgue o livro pela capa! O design pode não ser de outro mundo, mas o código é digno do Multiverso! ✨🚀/i
    );
  });
  it("Deveria ser possível buscar os personagens. Ao buscar deve redirecionar para outra tela", () => {
    jest.clearAllMocks();
    const valueSearch = "hulk";
    const pathSearchCharacters = "/";

    const searchSucess = "Buscou com sucesso";
    const screen = render(
      <RoutersContextTesting
        routes={[
          <Route
            path={`/characters`}
            element={<>{searchSucess}</>}
            key={valueSearch}
          />,
        ]}
      >
        <SearchCharacters />
      </RoutersContextTesting>
    );
    const SearchCharactersPageSearchBar = screen.getByTestId(
      "id-search-characters-bar"
    );
    const SearchCharactersPageSearchButton = screen.getByTestId(
      "id-search-characters-button"
    );

    expect(window.location.pathname).toBe(pathSearchCharacters);

    fireEvent.change(SearchCharactersPageSearchBar, {
      target: { value: valueSearch },
    });

    expect(SearchCharactersPageSearchBar).toHaveValue(valueSearch);

    fireEvent.click(SearchCharactersPageSearchButton);
    expect(window.location.pathname).not.toBe(pathSearchCharacters);
    expect(screen.container).toContainHTML(searchSucess);
    expect(screen.container).toMatchSnapshot();
    expect.assertions(5);
  });
  it("Deveria ao pesquisar e clicar no Botão de pesquisa navegar para a URL dos personagens com o inicial ou nome dos personagens por parametro", async () => {
    const valueSearch = "Hu";
    const pathCharacters = "/characters";
    const pathCharacterWithParamNameStarts =
      pathCharacters + "?nameStartsWith=" + valueSearch;

    const screen = render(
      <RoutersContextTesting
        routes={[
          <Route
            path={pathCharacters}
            element={<div>Characters Page 1</div>}
            key={1}
          />,
        ]}
      >
        <SearchCharacters />
      </RoutersContextTesting>
    );

    const SearchCharactersPageSearchBar = screen.getByTestId(
      "id-search-characters-bar"
    );
    const SearchCharactersPageSearchButton = screen.getByTestId(
      "id-search-characters-button"
    );

    expect(window.location.href).not.toMatch(pathCharacterWithParamNameStarts);

    fireEvent.change(SearchCharactersPageSearchBar, {
      target: { value: valueSearch },
    });

    expect(SearchCharactersPageSearchBar).toHaveValue(valueSearch);

    fireEvent.click(SearchCharactersPageSearchButton);

    expect(window.location.href).toMatch(pathCharacterWithParamNameStarts);
    expect(screen.container).toMatchSnapshot();
    expect.assertions(4);
  });
  it("Deveria ao pesquisar e pressionar a tecla Enter fazer a navegação para a URL dos personagens", () => {
    const valueSearch = "Hu";
    const pathCharacters = "/characters";
    const pathCharacterWithParamNameStarts =
      pathCharacters + "?nameStartsWith=" + valueSearch;

    const screen = render(
      <RoutersContextTesting
        routes={[
          <Route
            path={pathCharacters}
            element={<div>Characters Page 2</div>}
            key={pathCharacters}
          />,
        ]}
      >
        <SearchCharacters />
      </RoutersContextTesting>
    );

    const SearchCharactersPageSearchBar = screen.getByTestId(
      "id-search-characters-bar"
    );

    expect(window.location.href).not.toMatch(pathCharacterWithParamNameStarts);

    fireEvent.change(SearchCharactersPageSearchBar, {
      target: { value: valueSearch },
    });

    expect(SearchCharactersPageSearchBar).toHaveValue(valueSearch);

    fireEvent.keyDown(SearchCharactersPageSearchBar, { key: "Enter", code: "Enter" });

    expect(window.location.href).toMatch(pathCharacterWithParamNameStarts);
    expect(screen.container).toMatchSnapshot();
    expect.assertions(4);
  });
});
