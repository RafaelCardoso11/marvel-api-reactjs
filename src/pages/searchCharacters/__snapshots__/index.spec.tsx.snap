// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchCharacters /> Deveria ao pesquisar e clicar no Bot√£o de pesquisa navegar para a URL dos personagens com o inicial ou nome dos personagens por parametro 1`] = `
<div>
  <div>
    Characters Page 1
  </div>
</div>
`;

exports[`<SearchCharacters /> Deveria ao pesquisar e pressionar a tecla Enter fazer a navega√ß√£o para a URL dos personagens 1`] = `
<div>
  <div>
    Characters Page 2
  </div>
</div>
`;

exports[`<SearchCharacters /> Deveria renderizar na tela os containers SearchCharactersPage, SearchCharactersPageTitle, SearchCharactersTitleIcon, SearchCharactersPageSearchContainer, SearchCharactersPageFooter 1`] = `
<div>
  <div
    class="sc-kvAqsT dwecmR"
    data-testid="id-search-characters-page"
  >
    <h1
      class="sc-grJzXs fhXCmH"
      data-testid="id-search-characters-title"
    >
      Search for your favorite hero
      <p
        class="sc-hpvRAK imfqkc"
        data-testid="id-search-characters-title-icon-container"
      >
        <svg
          data-testid="id-search-characters-title-icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
          />
        </svg>
      </p>
    </h1>
    <h2
      class="sc-cdtKKh eLXtmM"
      data-testid="id-search-characters-subtitle"
    >
      Find information about your favorite Marvel heroes.
    </h2>
    <div
      class="sc-jWWaec fVdTEg"
      data-testid="id-search-characters-container"
    >
      <input
        class="sc-gZGWnS ckdMuw"
        data-testid="id-search-characters-bar"
        placeholder="Search for your favorite Marvel Hero"
        type="text"
        value=""
      />
      <button
        class="sc-cLrhaH fhEwuW"
        data-testid="id-search-characters-button"
      >
        Search
      </button>
    </div>
    <p
      class="sc-fIGkQR cnrkPR"
      data-testid="id-search-characters-page-footer"
    >
      ‚ö°Ô∏è N√£o julgue o livro pela capa! O design pode n√£o ser de outro mundo, mas o c√≥digo √© digno do Multiverso! ‚ú®üöÄ
    </p>
  </div>
</div>
`;

exports[`<SearchCharacters /> Deveria ser poss√≠vel buscar os personagens. Ao buscar deve redirecionar para outra tela 1`] = `
<div>
  Buscou com sucesso
</div>
`;
